pySDC/implementations/problem_classes/generic_spectral.py

```diff
@@ -0,0 +1,357 @@
+        comm=None,
+        Dirichlet_recombination=True,
+        left_preconditioner=True,
+        solver_type='cached_direct',
+        solver_args=None,
+        useGPU=False,
+        max_cached_factorizations=12,
+        spectral_space=True,
+        real_spectral_coefficients=False,
+        *args,
+        **kwargs,
+    ):
+        """
+        Base class for problems discretized with spectral methods.
+
+        Args:
+            bases (list of dictionaries): 1D Bases
+            components (list of strings): Components of the equations
+            comm (mpi4py.Intracomm or None): MPI communicator
+            Dirichlet_recombination (bool): Use Dirichlet recombination in the last axis as right preconditioner
+            left_preconditioner (bool): Reverse the Kronecker product if yes
+            solver_type (str): Solver for linear systems
+            solver_args (dict): Arguments for linear solver
+            useGPU (bool): Run on GPU or CPU
+            max_cached_factorizations (int): Number of matrix decompositions to cache before starting eviction
+            spectral_space (bool): If yes, the solution will not be transformed back after solving and evaluating the RHS, and is expected as input in spectral space to these functions
+            real_spectral_coefficients (bool): If yes, allow only real values in spectral space, otherwise, allow complex.
+        """
+        solver_args = {} if solver_args is None else solver_args
+        self._makeAttributeAndRegister(
+            'max_cached_factorizations',
+            'useGPU',
+            'solver_type',
+            'solver_args',
+            'left_preconditioner',
+            'Dirichlet_recombination',
+            'comm',
+            'spectral_space',
+            'real_spectral_coefficients',
+            localVars=locals(),
+        )
+        self.spectral = SpectralHelper(comm=comm, useGPU=useGPU)
+
+        if useGPU:
+            self.setup_GPU()
+
+        for base in bases:
+            self.spectral.add_axis(**base)
+        self.spectral.add_component(components)
+
+        self.spectral.setup_fft(real_spectral_coefficients)
+
+        super().__init__(init=self.spectral.init_forward if spectral_space else self.spectral.init)
+
+        self.work_counters[solver_type] = WorkCounter()
+        self.work_counters['factorizations'] = WorkCounter()
+
+        self.setup_preconditioner(Dirichlet_recombination, left_preconditioner)
+
+        self.cached_factorizations = {}
+
+    def __getattr__(self, name):
+        return getattr(self.spectral, name)
+
+    def _setup_operator(self, LHS):
+        """
+        Setup a sparse linear operator by adding relationships. See documentation for ``GenericSpectralLinear.setup_L`` to learn more.
+
+        Args:
+            LHS (dict): Equations to be added to the operator
+
+        Returns:
+            sparse linear operator
+        """
+        operator = self.spectral.get_empty_operator_matrix()
```

Comment:
... especially since you are using `self.spectral.get_[...]` here and not `self.get_[...]` which you could theoretically do ?

Reply:
In this class I always use `self.spectral.[...]`, but if you look in the actual problem implementations, you will find I don't do it there, which makes it a lot more readable. The goal is to do a Dedalus-like thing where users don't need to worry so much. Of course, they need to know what they are doing here a lot more. I don't claim to be competitive with Dedalus. But I think this is fine.